### ISSUE TITLE
FastFMIndexConfig problems and Function to build index

### ISSUE THREAD

**Issue by xhongyi** [seqan/seqan#2087]
Hi I have 2 separate questions:
1. It seems that for some reason, if I use FastFMIndexConfig instead of FMIndexConfig the finder won't find the correct query sequence. Is there special settings that I missed?

2. Is there anyway to load the Text beyond calling the constructor? For example, I have to StringSets but I only want to use a single FMIndex, can I just unload the previous and load the next StringSet? I can't find the proper interface in the API doc.

**Comment by cpockrandt** [comment]
Hi Hongyi,

1. Can you provide an example with data? I wasn't able to reproduce this behavior in a quick test.

2. I would suggest to move your code into a separate function and call it twice (once for each StringSet). This way you would essentially only have one FMIndex. We don't officially support modifying the text in an FMIndex and rebuilding it.

**Comment by xhongyi** [comment]
Sure. Try the code below. If I swap out "typedef FastFMIndexConfig<void, uint64_t> TConfig;" to "typedef FMIndexConfig<void, uint64_t> TConfig;", it prints "<0, 0>". Otherwise it prints nothing.

#include <iostream>
#include <seqan/sequence.h>
#include <seqan/index.h>

using namespace seqan;

typedef FastFMIndexConfig<void, uint64_t> TConfig;
typedef Index<StringSet<String<Dna5> >, FMIndex<void, TConfig> > BWTIndex;

BWTIndex fmIndex;
Finder<BWTIndex> fmFinder;

int main()
{
    StringSet<String<Dna5> > ref;
    appendValue (ref, "GTGCTCATCTCCTTGGCTGTGATACGTGGCCGGCCCTCGCTCCAGCAGCTGGACCCCTAC");

    fmIndex = BWTIndex(ref);
    setHaystack(fmFinder, fmIndex);
    while(find(fmFinder, "GTGCTCATCTCCTTGGCTGTGATACGTGGCCGGCCCTCGCTCCAGCAGCTGGACCCCTAC"))
         std::cout << position(fmFinder) << std::endl;
    return 0;
}

**Comment by xhongyi** [comment]
This leads to my other point. I was trying to use the Iterator instead of Finder. However, I always run into some problems with it.

For example, after the above code, if I use:

`Iter<BWTIndex, VSTree<TopDown<ParentLinks<> > > > bwt_iter(fmIndex);`
`goDown(bwt_iter, String<Dna5>("TCATCTCCTTGGC") );`

Then I will run into a segfault.

The part of index iterator and all the associated tags is not very comprehensively documented and it is hard to follow them in the API document.

**Comment by xhongyi** [comment]
Hi, it has been a while. Can I ask if the problem has been confirmed? Or I missed something with FastFMIndexConfig?

**Comment by cpockrandt** [comment]
Hi @xhongyi 

Sorry for coming back so late to you. I'm still looking into this issue. Insert a SeqAn-String into the find-method, then it should work:

    while(find(fmFinder, Dna5String("GTGCTCATCTCCTTGGCTGTGATACGTGGCCGGCCCTCGCTCCAGCAGCTGGACCCCTAC")))

But I agree that this behavior is odd and should be the same for both configurations. I'll let you know when I made a PR to fix this.

Regarding the other issue: can you provide a code sample again? I couldn't reproduce a segfault. Thanks!

**Comment by xhongyi** [comment]
@cpockrandt Actually, it worked.

However, there is a still a catch. If I load in a large dummy sequence, in my case the partial first chromosome of human_genome_v37.fasta (chr1). Build the FMIndex of that chromosome (with StringSet<String<Dna5> >). Then if I use the following code:

   while(find(fmFinder, String<Dna5>("GGTGGGCAGTTCTGGAATGGTGCCAGGGGCAGAGGGGGCAATGCCGGGGCCCAGGTC")))
        std::cout << position(fmFinder) << std::endl;

    std::cout << "-----" << std::endl;

    Iter<BWTIndex, VSTree<TopDown<ParentLinks<> > > > bwt_iter(fmIndex);
    goDown(bwt_iter, String<Dna5>("GGTGGGCAGTTCTGGAATGGTGCCAGGGGCAGAGGGGG") );

    std::cout << "Done!" << std::endl;

Then it would only print out

< 0 , 16887 >
-----
Done!

which means, the Iterator failed. Since "GGTGGGCAGTTCTGGAATGGTGCCAGGGGCAGAGGGGG" is only a prefix of "GGTGGGCAGTTCTGGAATGGTGCCAGGGGCAGAGGGGGCAATGCCGGGGCCCAGGTC". It can't be that the fmIndex can find "GGTGGGCAGTTCTGGAATGGTGCCAGGGGCAGAGGGGGCAATGCCGGGGCCCAGGTC" but not "GGTGGGCAGTTCTGGAATGGTGCCAGGGGCAGAGGGGG".

**Comment by xhongyi** [comment]
@cpockrandt The same issue won't happen for a smaller sequence.

**Comment by xhongyi** [comment]
@cpockrandt is there a way I can send you code snippets?

Iterator with FMIndex has many very weird behaviors. Depending on the StringSet, it is sometimes correct and other times very incorrect. It is as if it can only travel at most 24 layers in the suffix tree before it stops.

It would be much easier if I can just send over code snippets so that you can easily reproduce the problems.

**Comment by xhongyi** [comment]
Chris pointed out in an email. Since forward BWT and FM can only be traversed in reverse fashion, in order to iterate through the FMIndex, the query sequence has to be reversed first.

Confirmed that this will fixed the problem. Ticket closed.
